<img src="../images/cads-logo.png" style="height: 100px;" align=left> <img src="../images/python-logo.png" style="height: 100px;" align=right>

# Python Fundamentals Day 3

Welcome to Python Fundamentals Day 3.

So what is on the menu for today? 

We start with reading and writing from files, learn how to transport data objects in the JSON format and how to browse the filesystem.

Then we move to Exception handling. We learn about the different types of Exceptions, what they mean and how to resolve them. Then we learn how to handle Exceptions in our code using the try-except statement.

Last but not least, we learn about Object Oriented Programming including class inheritance and how to define and use your own classes in your coding.

Let's go!

## Table of Contents

- [File I/O](#File-I/O)
    - [Reading and writing lines with open](#Reading-and-writing-lines-with-open)
    - [JSON](#JSON)
    - [File browsing with glob](#File-browsing-with-glob)
    - [Summary](#Summary)
        - [open](#open)
        - [JSON](#JSON)
        - [import](#import)
        - [glob](#glob)
    - [RUN ME](#RUN-ME)
    - [Exercises](#Exercises)
        - [Exercise 1 - Read in stocks](#Exercise-1---Read-in-stocks)
        - [Exercise 2 - First ten names](#Exercise-2---First-ten-names)
        - [Exercise 3 - Inc only](#Exercise-3---Inc-only)
        - [Exercise 4 - Average PE](#Exercise-4---Average-PE)
- [Exception Handling](#Exception-Handling)
    - [Exceptions](#Exceptions)
    - [Try except](#Try-except)
    - [Summary](#Summary)
        - [try except](#try-except)
        - [try except else finally](#try-except-else-finally)
    - [RUN ME](#RUN-ME)
    - [Exercises](#Exercises)
        - [Exercise 1 - Fix it Multiply](#Exercise-1---Fix-it-Multiply)
        - [Exercise 2 - Fix it Numbers](#Exercise-2---Fix-it-Numbers)
        - [Exercise 3 - Fix it Open](#Exercise-3---Fix-it-Open)
        - [Exercise 4 - Try salaries](#Exercise-4---Try-salaries)
- [Object Oriented Programming](#Object-Oriented-Programming)
    - [Summary](#Summary)
        - [class](#class)
    - [RUN ME](#RUN-ME)
    - [Exercises](#Exercises)
        - [Exercise 1 - Get pokemon info](#Exercise-1---Get-pokemon-info)
        - [Exercise 2 - Create a pokemon class](#Exercise-2---Create-a-pokemon-class)
        - [Exercise 3 -  Load all pokemons](#Exercise-3----Load-all-pokemons)
        - [Exercise 4 - Provide insights](#Exercise-4---Provide-insights)

## File I/O

### Reading and writing lines with open

Communicating with files is quite useful in programming.

With the open() built-in function we can read, write and append to files.

First check what is your working directory. You will be able to find the file that we will read and write there.

In [1]:
%pwd # tell me where I am currently now --> magic function

'C:\\Users\\Acer\\Desktop\\Data Analytics\\Python for Analytics (Basic)\\Python for Analytics (Basic)\\Day 3'

We now create a file object which we will use to write something to file.

In [5]:
f = open('test.txt','w') # declare file name, 'w' = write dalam folder yg sama
f.write('Hey haseena') # calling it to open then write 'Hey' in the text file
# output kluar how many character dalam tu

11

Now check the file in your folder. Can you see Hey! in test.txt? 

No you cannot. Yet..

In [6]:
f.close() # kena close dulu baru text file tu update

Check test.txt again. 

Now it is there.

So it is important to not forget to close the file object.

There is an elegant construct for this. Let's try it.

In [7]:
# with open--> x payah close()

with open ('test2.txt','w') as f: # write akan overwrite text lama
    f.write('Hello there')

Check again. Test test2.txt file is there.

So what did we do here?

open() returns a file object and it is commonly used with two arguments: open(file, mode)

- file is a string containing the filename
- mode describes in which way the file will be used: writing, reading, appending.

mode can be:
- `'r'` read (by default)
- `'w'` write
- `'a'` append

Notice the fat-green with as statement. Within the with statement the file object will be available as f. After the statement the file object will be closed. 

In [11]:
with open('test2.txt', 'r')  as f: # f tu point dkat file, boleh letak huruf lain
    print(f.read()) # baca apa yg ada kat file tu

Hello there


In [20]:
with open('test2.txt', 'a') as f: # nak tambah line dlm file tu
    f.write('\n')
    f.write('Hellloooooo')
    f.write('\nHellloooooo2')
    f.write('\nHellloooooo3')

In [21]:
with open('test2.txt', 'r')  as f: # f tu point dkat file, boleh letak huruf lain
    print(f.read())

Hello there
Hellloooooo
Hellloooooo
Hellloooooo2
Hellloooooo
Hellloooooo
Hellloooooo2
Hellloooooo3
Hellloooooo
Hellloooooo2
Hellloooooo3


Now we can check the contents of the file using 

!cat -- for Mac / Linux users

!type -- for Windows users

In [19]:
!type test2.txt # boleh pakai ni jugak utk read file

Hello there
Hellloooooo
Hellloooooo
Hellloooooo2
Hellloooooo
Hellloooooo
Hellloooooo2
Hellloooooo3



test2.txt


The system cannot find the file specified.
Error occurred while processing: #.
The system cannot find the file specified.
Error occurred while processing: boleh.
The system cannot find the file specified.
Error occurred while processing: pakai.
The system cannot find the file specified.
Error occurred while processing: ni.
The system cannot find the file specified.
Error occurred while processing: jugak.
The system cannot find the file specified.
Error occurred while processing: utk.
The system cannot find the file specified.
Error occurred while processing: read.
The system cannot find the file specified.
Error occurred while processing: file.


Use **strip()** to remove all whitespace characters from the beginning and the end of the string.

In [24]:
text = """
The names are:

Jeremy
Narjes
Amin
""".strip() # buang white space je, /n still ada

text

'The names are:\n\nJeremy\nNarjes\nAmin'

In [23]:
text = """
The names are:

Jeremy
Narjes
Amin
""".strip().replace('\n','')

text

'The names are:JeremyNarjesAmin'

In [25]:
import os # nak tau directory file, bagitau file tu ada kat mana sbb xde dlam current directory 
file_names = os.path.join("..","data", "names_raw.txt")
file_names
# .. = 2 folders back

'..\\data\\names_raw.txt'

The import statement allows you to import libraries which you can use. 

Here we use os.path.join to safely join the paths of the data folder and the names_raw. txt file so that it works for any operating system (Windows/Mac).

You can iterate over the file object and do something line by line.

In [26]:
# read file , nak output line by line
with open(file_names, 'r') as f:
    for line in f:
        print(line)

The names are:



Jeremy

Narjes

Amin


Let's say now we want to transform this file into a list of names. 

**Question:** How can we transform this into a list of the names?


Let's parse the contents of this file into a list of names.

In [31]:
# put the line by line into list
with open(file_names, 'r') as f:
    lines_list = [] # buant container kosong
    for line in f:
        lines_list.append(line.strip()) # line.strip utk buang semua whitespace
        # masukkan dalam container kosong tu
print(lines_list[:]) # slicing
print(lines_list[2:])

['The names are:', '', 'Jeremy', 'Narjes', 'Amin']
['Jeremy', 'Narjes', 'Amin']


### JSON

JSON (JavaScript Object Notation) is an open-standard format that uses human-readable text to transmit data objects consisting of attribute–value pairs. 

It is used often by API's (Application Programming Interface) to communicate data between programs or to commumnicate data between an applications and users.

The notation is almost the same as the dictionary in Python except for:

- JSON: true/false -- Python: True/False
- JSON: null -- Python: None

Here is an example JSON from [wikipedia](https://en.wikipedia.org/wiki/JSON#Data_types.2C_syntax_and_example)

```json
{
  "firstName": "John",
  "lastName": "Smith",
  "isAlive": true,
  "age": 25,
  "address": {
    "streetAddress": "21 2nd Street",
    "city": "New York",
    "state": "NY",
    "postalCode": "10021-3100"
  },
  "phoneNumbers": [
    {
      "type": "home",
      "number": "212 555-1234"
    },
    {
      "type": "office",
      "number": "646 555-4567"
    },
    {
      "type": "mobile",
      "number": "123 456-7890"
    }
  ],
  "children": [],
  "spouse": null
}
```

First we import the json library to provide us with the functionality. 

In [34]:
import json

We can transform a data object into a json string and transform it back. 

In [32]:
data = {"name": "Jeremy", "grades": [10, 15], "teacher":True, "complaints":None}
data #dict

{'name': 'Jeremy', 'grades': [10, 15], 'teacher': True, 'complaints': None}

In [35]:
# use dumps to save dictionary into json format
# convert data in the dict into json format
data_json = json.dumps(data)
data_json

'{"name": "Jeremy", "grades": [10, 15], "teacher": true, "complaints": null}'

**Question:** Do you see the difference between a dictionary and a JSON?
<br><br><br>
**Notice the difference in <u>true</u> and <u>null</u>**. Also notice <u>the addidtional quotes at the start and the end as this is a string.</u>

Now we del data. Then load it again from the JSON file.

In [ ]:
# format
# dict: none  json: null
# dict: True  json: true

In [36]:
del data  # delete data

In [37]:
data # dah xde data dalam version dictionary

NameError: name 'data' is not defined

In [38]:
data = json.loads(data_json) # version json still ada
data # so kita reload data json dalam dictionary

{'name': 'Jeremy', 'grades': [10, 15], 'teacher': True, 'complaints': None}

In [39]:
data['grades'] # nak tau grades 

[10, 15]

Now we can dump and load data from text files that contain json directly.

In [40]:
# create variable 'data' with value in dict format
data = [{'name': 'Jeremy', 'grades': [10, 15], 'teacher': True, 'complaints': None},
        {'name': 'Amin', 'grades': [3, 4], 'teacher': True, 'complaints': "Many"},
        {'name': 'Narjes', 'grades': [8, 13, 20], 'teacher': True, 'complaints': None},
        {'name': 'Laleh', 'grades': [5, 5, 5], 'teacher': False, 'complaints': "Many, so many!"}
       ]

In [41]:
with open('cads.json', 'w') as f: # create json file
    json.dump(data,f) # dump(convert) the dict data into the json file
    
# json,csv file is more efficient in data analysis

Let's look at the resulting json file using !cat for Mac/Linux and !type for windows

In [42]:
!type cads.json # a json file with json format created

[{"name": "Jeremy", "grades": [10, 15], "teacher": true, "complaints": null}, {"name": "Amin", "grades": [3, 4], "teacher": true, "complaints": "Many"}, {"name": "Narjes", "grades": [8, 13, 20], "teacher": true, "complaints": null}, {"name": "Laleh", "grades": [5, 5, 5], "teacher": false, "complaints": "Many, so many!"}]


Now we can load that data back into a dictionary

EXERCISE - MAKE YOUR OWN DICT INTO JSON FILE

In [44]:
# Haseena
New_data = [{'Name':'Aiman','Product':'Syampoo','Price':4.50, 'Complaint': 40},
           {'Name':'Aisya','Product':'Conditioner','Price':34,'Complaint': None},
           {'Name':'Haseena','Product':'Hair brush','Price':7.56, 'Complaint':3},
           {'Name':'Ali', 'Product':'Scisor','Price':39,'Complaint':None}]

New_data

[{'Name': 'Aiman', 'Product': 'Syampoo', 'Price': 4.5, 'Complaint': 40},
 {'Name': 'Aisya', 'Product': 'Conditioner', 'Price': 34, 'Complaint': None},
 {'Name': 'Haseena', 'Product': 'Hair brush', 'Price': 7.56, 'Complaint': 3},
 {'Name': 'Ali', 'Product': 'Scisor', 'Price': 39, 'Complaint': None}]

In [45]:
with open('shop.json','w') as f:
    json.dump(New_data,f)

In [46]:
!type shop.json

[{"Name": "Aiman", "Product": "Syampoo", "Price": 4.5, "Complaint": 40}, {"Name": "Aisya", "Product": "Conditioner", "Price": 34, "Complaint": null}, {"Name": "Haseena", "Product": "Hair brush", "Price": 7.56, "Complaint": 3}, {"Name": "Ali", "Product": "Scisor", "Price": 39, "Complaint": null}]


### File browsing with glob

With glob you can easily list files and folders in a directory

In [47]:
import glob # bagi list brapa text file dlm directory skarang

folder = os.path.join("..", "data", "folders")

for filename in glob.glob('{}/*.txt'.format(folder)):
    print(filename)

..\data\folders\base_one_file.txt
..\data\folders\base_two_file.txt


And recursively.

You can get the files in a list like this.

In [48]:
import glob
# ada text file apa dalam folder 'data'
folder = os.path.join("..", "data", "folders") 
for filename in glob.glob('{}/**/*.txt'.format(folder), recursive=True):
    print(filename)
    

# recursive=True: it will check every single folder that we have in that file

..\data\folders\base_one_file.txt
..\data\folders\base_two_file.txt
..\data\folders\folder_1\one_file.txt
..\data\folders\folder_1\two_file.txt
..\data\folders\folder_2\three_file.txt
..\data\folders\folder_3\five_file.txt
..\data\folders\folder_3\four_file.txt


### Summary

> #### open
Built-in function to open a file object to read/write/append to a file.
```python
with open('test.txt', 'w') as f:
    f.write("Hello")
with open('test2.txt', 'r') as f:
    data = f.readlines()
with open('test2.txt', 'r') as f:
    f.readline()
    data = []
    for line in f:
        data.append(line)
```

> #### JSON
JSON (JavaScript Object Notation) is an open-standard format that uses human-readable text to transmit data objects consisting of attribute–value pairs. Used often to communicate data with API's.
```python
import json
json.dumps(data)
json.loads(json_str)
with open('data.json', 'w') as f:
    json.dump(data, f)
with open('data.json', 'r') as f:
    data = json.load(f)
```

> #### import
import libraries to use the functionality
```python
import json
import os
```

> #### glob
module that is used for file listing
```python
for filename in glob.glob('{}/*.txt'.format(folder_path)):
    print(filename)
for filename in glob.glob('{}/**/*.txt'.format(folder_path), recursive=True):
    print(filename)
```

### RUN ME

Please run the below code snippet. It is required for running tests for your solution.

In [49]:
def test(got, expected):
    if got == expected:
        prefix = ' OK '
    else:
        prefix = '  FAIL '
    print((f' {prefix} got: {got} expected: {expected}'))

In [50]:
def test(got, expected):
    max_width_print = min([int(79/3), 
                           len(repr(got)) + 10, 
                           len(repr(expected)) + 10])
    
    if got == expected:
        prefix = ' OK '.center(max_width_print, '=')
    else:
        prefix = ' FAIL '.center(max_width_print, '=')
    print(('%s\n     got: %s \nexpected: %s' % (prefix, repr(got), repr(expected))), 
          end = '\n\n')


In [51]:
test('a', 'ab')
test('a', 'a')

==== FAIL ===
     got: 'a' 
expected: 'ab'

===== OK ====
     got: 'a' 
expected: 'a'



### Exercises

#### Exercise 1 - Read in stocks

Read in stocks data in "data/stocks.json" and store it into stocks.

Each line of this file is a json object, but the whole file content is not a valid json object since the jsons are no separated by a comma and there is no extra brackets surrounding the jsons.

So you need to read the file line by line and transform those strings into dict using json.loads.

Store the dictionaries into stocks.

Hints: import os, os.path.join, import json, json.loads

In [ ]:
with open('stocks.json', 'r') as f:
    data = json.load(f)
        
# TEST
print("read_in_stocks")
test(len(stocks), 6756)
test(type(stocks[0]), dict)

In [53]:
import os
import json

stocks_file = os.path.join('..','data','stocks.json')
stocks = []
with open(stocks_file, 'r') as f:
    for line in f:
        stocks.append(json.loads(line)) # load into dict, then append to list line by line
        
print("read_in_stocks")
test(len(stocks), 6756)
test(type(stocks[0]), dict)

read_in_stocks
===== OK =====
     got: 6756 
expected: 6756

========== OK ==========
     got: <class 'dict'> 
expected: <class 'dict'>



#### Exercise 2 - First ten names

What are the names of the first ten companies?

Hints: inspect one dict to see which key contains the name.

In [ ]:
names = ## your code

# TEST
print("first_ten_names")
from answers import ten_companies
test(names, ten_companies)

In [70]:
names = [stock['Company'] for stock in stocks[:10]]

In [64]:
for stock in stocks: # listkan semua
    name = stock['Company']
    print(name)


Agilent Technologies Inc.
Alcoa, Inc.
WCM/BNY Mellon Focused Growth ADR ETF
iShares MSCI AC Asia Information Tech
Altisource Asset Management Corporation
Atlantic American Corp.
Aaron's, Inc.
Applied Optoelectronics, Inc.
AAON Inc.
Advance Auto Parts Inc.
Apple Inc.
American Assets Trust, Inc.
Almaden Minerals Ltd.
Advantage Oil & Gas Ltd.
Atlas Air Worldwide Holdings Inc.
iShares MSCI All Country Asia ex Jpn Idx
AllianceBernstein Holding L.P.
Abaxis Inc.
ABB Ltd.
AbbVie Inc.
AmerisourceBergen Corporation
Ameris Bancorp
Cambium Learning Group, Inc.
Advisory Board Co.
Arkansas Best Corporation
Asbury Automotive Group, Inc.
ARCA biopharma, Inc.
ABM Industries Inc.
American Bio Medica Corp.
Abiomed Inc.
Arbor Realty Trust Inc.
Abbott Laboratories
Autobytel Inc.
Companhia de Bebidas Das Americas (AMBEV)
Barrick Gold Corporation
ACADIA Pharmaceuticals, Inc.
American Capital, Ltd.
Arctic Cat Inc.
American Campus Communities Inc.
Accelrys Inc.
ACCO Brands Corporation
AdvisorShares Accuvest Gl

Cellcom Israel Ltd.
Celgene Corporation
Celsius Holdings, Inc.
China Electric Motor, Inc.
Chembio Diagnostics, Inc.
Cempra, Inc.
Central Garden & Pet Co.
Century Aluminum Co.
CNOOC Ltd.
Constellation Energy Partners LLC
Crestwood Equity Partners LP
Ceres, Inc.
Cerner Corporation
Cereplast, Inc.
Cerus Corporation
Central Securities Corp.
Central European Media Enterprises Ltd.
Eaton Vance California Municipal Income Trust
CEVA Inc.
WisdomTree Dreyfus Emerging Currency
CF Industries Holdings, Inc.
Central Federal Corp.
C&F Financial Corp.
Capitol Federal Financial, Inc.
Culp Inc.
CareFusion Corporation
California First National Bancorp
Cardinal Financial Corp.
Cornerstone Progressive Return 
Cullen/Frost Bankers, Inc.
iShares Barclays Credit Bond
Colfax Corporation
The Carlyle Group LP
China Green Agriculture, Inc.
Compugen Ltd.
CGG SA
Celadon Group Inc.
Cognex Corporation
Calamos Global Total Return Fund
Claude Resources, Inc.
Guggenheim S&P Global Water Index
Consolidated Graphics, Inc

db X-trackers MSCI UK Hdgd Eq ETF
PowerShares DB G10 Currency Harvest
DCA Total Return Fund
Donaldson Company, Inc.
Digital Cinema Destinations Corp.
Diana Containerships Inc.
NTT DOCOMO, Inc.
iPath Seasonal Natural Gas ETN
Ducommun Inc.
Dime Community Bancshares Inc.
DCT Industrial Trust Inc.
Delcath Systems, Inc.
E. I. du Pont de Nemours and Company
Dominion Diamond Corporation
3D Systems Corp.
Dover Downs Gaming & Entertainment Inc.
Delaware Investments Dividend & Income Fund, Inc.
ProShares Short Oil & Gas
ProShares Ultra Dow30
PowerShares DB Commodity Short ETN
DDR Corp.
Dillard's Inc.
Deere & Company
Dearborn Bancorp Inc.
Deckers Outdoor Corp.
PowerShares DB Commodity Dble Short ETN
Deer Consumer Products, Inc.
Guggenheim Defensive Equity
Etablissements Delhaize Freres et Cie Le Lion S.A.
Douglas Emmett Inc
Dejour Energy Inc.
Deltic Timber Corp.
Dell Inc.
WisdomTree Emerging Markets Equity Inc
Denny's Corporation
Diageo plc
DepoMed Inc.
WisdomTree SmallCap Dividend
Destination Ma

Essex Rental Corp.
Earthstone Energy, Inc.
Ensco plc
Community Bankers Trust Corporation
Elecsys Corporation
Elephant Talk Communications Corp.
Eaton Vance Tax-Managed BuyWrite Income Fund
Energy Transfer Equity, L.P.
Aberdeen Emerging Markets Telecommunications Fund, Inc.
E*TRADE Financial Corporation
Eaton Vance Tax Advantaged Global Dividend Income Fund
Ethan Allen Interiors Inc.
Eaton Vance Risk-Managed Divers
Entercom Communications Corp.
Eaton Corporation
Eaton Vance Tax-Advantaged Global Dividend Opportunities Fund
Energy Transfer Partners LP
Entergy Corporation
EnteroMedics, Inc.
Eaton Vance Tax-Managed Buy-Write Opportunities Fund
Eaton Vance Tax-Managed Global Buy-Write Opportunities Fund
Eaton Vance Tax-Managed Diversified Equity Income Fund
WisdomTree Euro Debt
iShares MSCI Europe Financials Index
ProShares Short MSCI Emerging Markets
ProShares UltraShort Euro
iShares MSCI USA Index
Eaton Vance Corp.
Edwards Group Limited
iShares MSCI Emerging Markets Value
Evans Bancorp In

Market Vectors Glb Alternatve Energy ETF
New Germany Fund Inc.
Gafisa S.A.
Guaranty Federal Bancshares Inc.
Griffon Corporation
Gold Fields Ltd.
GFI Group Inc.
General Finance Corporation
Western Asset Variable Rate Strategic Fund Inc.
Goldcorp Inc.
Grupo Financiero Galicia S.A.
Gerdau S.A.
Graco Inc.
The Gabelli Global Gold, Natural Resources & Income Trust
ProShares German Sov / Sub-Sov Debt ETF
General Growth Properties Inc.
Global Geophysical Services, Inc.
Gabelli Global Multimedia Trust Inc.
Genomic Health Inc.
Global High Income Fund, Inc.
Greenhill & Co., Inc.
Graham Corp.
iShares Global High Yield Corporate Bond
CGI Group, Inc.
Gulf Island Fabrication Inc.
GigOptix, Inc.
Giga-tronics Inc.
GigaMedia Ltd.
SPDR FTSE/Macquarie Global Infra 100
G-III Apparel Group, Ltd.
Gildan Activewear Inc.
Gilead Sciences Inc.
Gilat Satellite Networks Ltd.
Templeton Global Income Fund, Inc.
Gigamon Inc.
General Mills, Inc.
Given Imaging Ltd.
Guggenheim Enhanced Core Bond
G&K Services Inc.
Geekne

iShares Dow Jones Intl Select Div Idx
Market Vectors Indonesia Index ETF
Market Vectors Indonesia Small-Cap ETF
IDEXX Laboratories, Inc.
IEC Electronics Corp.
iShares Barclays 7-10 Year Treasury
iShares Barclays 3-7 Year Treasury Bond
iShares Enhanced US Large-Cap
iShares Dow Jones US Oil & Gas Ex Index
Icahn Enterprises, L.P.
Integrated Electrical Services, Inc.
iShares Enhanced US Small-Cap
iShares S&P Europe 350 Index
IDEX Corporation
iShares Dow Jones US Oil Equipment Index
Aberdeen Indonesia Fund, Inc.
iShares FTSE EPRA/NAREIT Dev Asia Idx
iShares FTSE EPRA/NAREIT Dev EU Idx
International Flavors & Fragrances Inc.
iShares FTSE EPRA/NAREIT Dev RE ex-US
Institutional Financial Markets, Inc.
India Fund, Inc.
iShares FTSE EPRA/NAREIT North America
Infinity Energy Resources Inc.
Infosonics Corp.
iShares FTSE Dev Sm Cap ex-North Amer
Imperial Holdings, Inc.
IGI, Laboratories, Inc.
ING Global Advantage and Premium Opportunity Fund
India Globalization Capital, Inc.
ING Global Equity Divid

Quaker Chemical Corporation
Market Vectors Solar Energy ETF
iShares S&P Global Consumer Staples
Kayne Anderson Energy Total Return Fund
Kayne Anderson MLP Investment Company
Kyocera Corp.
Kythera Biopharmaceuticals, Inc.
Loews Corporation
Louisiana Bancorp, Inc.
Multi-Color Corp.
Lakes Entertainment Inc.
Lithia Motors Inc.
SPDR Barclays Capital Aggregate Bond
Lakeland Industries Inc.
Lamar Advertising Co.
Lancaster Colony Corporation
Gladstone Land Corporation
Aberdeen Latin America Equity Fund, Inc.
Landmark Bancorp Inc.
Lentuo International Inc.
Market Vectors LatAm Small-Cap Index ETF
Lawson Products Inc.
Layne Christensen Co.
Lazard Ltd.
Lakeland Bancorp Inc.
DWS Global High Income Fund, Inc.
Leading Brands Inc.
Direxion Daily Latin America Bull 3X Shs
PowerShares DB 3x Lng 25+ Yr Trsy Bd ETN
Liberty Global Inc.
Libbey Inc.
LCA-Vision Inc.
US Airways Group, Inc.
Lannett Company, Inc.
Advent/Claymore Enhanced Growth & Income Fund
LCNB Corp.
Lifetime Brands, Inc.
iPath DJ-UBS Lead TR

LocalShares Nashville Area ETF
Pax MSCI North America ESG Index ETF
Nordic American Tankers Limited
Nathan's Famous Inc.
National Instruments Corporation
National Interstate Corporation
Nature's Sunshine Products Inc.
National American University Holdings, Inc.
Navistar International Corporation
Navidea Biopharmaceuticals, Inc
Navigators Group Inc.
Nuveen Arizona Premium Income Municipal Fund, Inc.
NewBridge Bancorp
First NBC Bank Holding Company
National Bank of Greece SA
Neuberger Berman Intermediate Municipal Fund Inc.
National Bank Holdings Corp.
Neurocrine Biosciences Inc.
Noble Energy, Inc.
Northeast Bancorp
Neuberger Berman New York Intermediate Municipal Fund Inc.
Nabors Industries Ltd.
Neostem, Inc.
NBT Bancorp, Inc.
NB&T Financial Group Inc.
Neuberger Berman California Intermediate Municipal Fund Inc.
NovaBay Pharmaceuticals, Inc.
Nacco Industries Inc.
Nuveen California Municipal Value Fund, Inc.
New Century Bancorp Inc.
Norcraft Companies, Inc. Common
Navigant Consulting Inc

Outerwall Inc.
OvaScience, Inc.
Ohio Valley Banc Corp.
Oak Valley Bancorp
Overland Storage Inc.
OmniVision Technologies Inc.
Orbitz Worldwide, Inc.
Oxygen Biotherapeutics, Inc.
Oxford Resource Partners, L.P.
OXiGENE, Inc.
Oxford Industries Inc.
Occidental Petroleum Corporation
Och-Ziff Capital Management Group LLC
Bank of the Ozarks, Inc.
Pandora Media, Inc.
Plains All American Pipeline, L.P.
Pan American Silver Corp.
Grupo Aeroportuario del Pacifico S.A.B. de CV
Pacific Biosciences of California, Inc.
Pacific Drilling S.A.
Pacer International Inc.
Pactera Technology International Ltd.
PacWest Bancorp
PowerShares FTSE RAFI Asia Pacific ex-Jp
Penske Automotive Group, Inc.
PowerShares Global Agriculture
Plains GP Holdings, L.P.
Western Asset Income Fund
North American Palladium Ltd.
ETFS Physical Palladium Shares
Pampa Energia SA
Parametric Sound Corp.
Palo Alto Networks, Inc.
PAR Technology Corp.
NuPathe, Inc.
Patrick Industries Inc.
Patriot Transportation Holding Inc.
VeriFone Systems,

Schwab U.S. Large-Cap Value ETF
The Charles Schwab Corporation
Schwab U.S. Large-Cap ETF
Schwab U.S. Aggregate Bond ETF
Service Corp. International
Market Vectors India Small-Cap ETF
Scientific Learning Corp.
EGShares India Small Cap
iShares MSCI Japan Small Cap Index
Socket Mobile, Inc.
Stepan Company
SciClone Pharmaceuticals, Inc.
Sucampo Pharmaceuticals, Inc.
Sycamore Networks Inc.
ProShares UltraShort DJ-UBS Crude Oil
SinoCoking and Coke Chemical Industries, Inc.
Superconductor Technologies Inc.
comScore, Inc.
SPDR Barclays Capital Short Term Corp Bd
Simcere Pharmaceutical Group.
Steelcase Inc.
ScanSource, Inc.
Select Comfort Corporation
SolarCity Corporation
Shoe Carnival Inc.
LS Starrett Co.
iShares MSCI EAFE Small Cap Index
SandRidge Energy, Inc.
ProShares UltraShort SmallCap600
Global X SuperDividend ETF
ProShares UltraShort Russell MCap Growth
Seadrill Partners LLC
ALPS Sector Dividend Dogs ETF
ProShares UltraPro Short Dow30
ProShares UltraShort Utilities
Sandridge Mississippi

TG Therapeutics, Inc.
Theragenics Corp.
Tenet Healthcare Corp.
iShares MSCI Thailand Invest Mkt Index
First Financial Corp.
The Hanover Insurance Group Inc.
Market Vectors Treasury-Hedged Hi Yld Bd
Tim Hortons Inc.
Threshold Pharmaceuticals Inc.
International Tower Hill Mines Ltd.
Thor Industries Inc.
Thoratec Corp.
Thermon Group Holdings, Inc.
TF Financial Corporation
Gentherm Incorporated
Theravance Inc.
Treehouse Foods, Inc.
Truett-Hurst, Inc.
THT Heat Transfer Technology, Inc.
Telecom Italia SpA
Telecom Italia SpA
TIBCO Software Inc.
TICC Capital Corp.
Tiffany & Co.
TigerLogic Corporation
Tel-Instrument Electronics Corp.
Interface Inc.
FlexShares Mstar US Mkt Factors Tilt ETF
Harris & Harris Group, Inc.
iShares Barclays TIPS Bond
Tiptree Financial Inc.
SPDR Barclays 1-10 Year TIPS ETF
PIMCO Broad U.S. TIPS Index ETF
Orchids Paper Products Company
Top Image Systems Ltd.
Team Inc.
Titan Machinery, Inc.
TiVo Inc.
The TJX Companies, Inc.
Teekay Corporation
Turkcell Iletisim Hizmetleri 

Walgreen Co.
WageWorks, Inc.
Wesco Aircraft Holdings, Inc.
Western Alliance Bancorporation
Washington Trust Bancorp Inc.
Waters Corp.
Wave Systems Corp.
Wayne Savings Bancshares Inc.
Westbury Bancorp, Inc.
WABCO Holdings Inc.
Washington Banking Co.
Westpac Banking Corporation
Wolverine Bancorp, Inc.
WebMD Health Corp.
Webster Financial Corp.
WESCO International Inc.
WellCare Health Plans, Inc.
Waste Connections Inc.
Walker & Dunlop, Inc.
Workday, Inc.
Western Digital Corp.
WD-40 Company
SPDR S&P Global Dividend ETF
Waddell & Reed Financial Inc.
WisdomTree Managed Futures
Western Asset Premier Bond Fund
Teucrium Wheat
Wellesley Bancorp, Inc.
Wisconsin Energy Corp.
iPath Pure Beta Grains ETN
The Wendy's Company
Werner Enterprises Inc.
Western Gas Partners LP
Westinghouse Solar, Inc.
WisdomTree Investments, Inc.
Winland Electronics Inc.
Weyco Group Inc.
Woori Finance Holdings Co., Ltd.
WashingtonFirst Bankshares Inc.
Wells Fargo & Company
Westfield Financial Inc.
Whole Foods Market, Inc.


#### Exercise 3 - Inc only

From the top 10 companies, now only show the names that contain the word 'Inc.'

In [ ]:
names = ## your code

# TEST
from answers import inc_companies
print("inc_only")
test(names, inc_companies)

In [71]:
name = [n for n in names if 'Inc' in n]

from answers import inc_companies
print("inc_only")
test(names, inc_companies)


inc_only
========== FAIL ==========
     got: ['Agilent Technologies Inc.', 'Alcoa, Inc.', 'WCM/BNY Mellon Focused Growth ADR ETF', 'iShares MSCI AC Asia Information Tech', 'Altisource Asset Management Corporation', 'Atlantic American Corp.', "Aaron's, Inc.", 'Applied Optoelectronics, Inc.', 'AAON Inc.', 'Advance Auto Parts Inc.'] 
expected: ['Agilent Technologies Inc.', 'Alcoa, Inc.', "Aaron's, Inc.", 'Applied Optoelectronics, Inc.', 'AAON Inc.', 'Advance Auto Parts Inc.']



#### Exercise 4 - Average PE

Now show the average P/E for all the data. Round it by 2.

Not all the stocks have the P/E reported, so you need to handle that. 

What percentage of the stocks has P/E reported? Round it by 2.

In [ ]:
## your code

avg_pe = ## your code
perc_with_pe = ## your code

# TEST
print("average_pe")
test(avg_pe, 41.71)
test(perc_with_pe, 0.5)

In [73]:
pe = [s['P/E'] for s in stocks if 'P/E' in s]
avg_pe = round(sum(pe) / len(pe),2)
perc_with_pe = round(len(pe) / len(stocks),2)

print("average_pe")
test(avg_pe, 41.71)
test(perc_with_pe, 0.5)

average_pe
====== OK =====
     got: 41.71 
expected: 41.71

===== OK ====
     got: 0.5 
expected: 0.5



## Exception Handling

### Exceptions

Here is an example Exception.

In [ ]:
while True print('Hello world')

- Arrow ^ pointing at the earliest point in the line where the error was detected
- The error is caused by the token preceding the arrow
- File name and line number are printed

There are different types of Exceptions.

Previous one was SyntaxError, other built-in exception types include:
- ZeroDivisionError -- division by zero
- NameError -- name not defined
- TypeError -- incorrect type
- KeyError -- key not found in dict
- IndexError -- index greater than length in list

The last line of the error message indicates what happened. Let's go over them.
Find more information [here](https://docs.python.org/3/library/exceptions.html).

In [75]:
10 * (1/0) # bahagi 0 = infinity --> ZeroDivisionError

ZeroDivisionError: division by zero

In [74]:
4 + result*3 # x declare lagi result --> NameError

NameError: name 'result' is not defined

In [77]:
'1' + 1 # string x boleh concatenate dgn integer --> TypeError

TypeError: can only concatenate str (not "int") to str

In [76]:
'1'+'1' # --> concantenate 2 string

'11'

In [78]:
d = {} # dictionary yang xde key
d[0] # x tau nak bagi output apa sbb xde key pon --> KeyError

KeyError: 0

In [79]:
l = [] # list kosong
l[0] # xde value dlm tu --> IndexError  

IndexError: list index out of range

The preceding part of the error message shows the context where the exception happened.

It contains a stack traceback listing source lines.

Here is an example.

In [80]:
def divide(x, y):
    return x/y    # 1 function utk 1 return kalau x declare variable

def call_divide(x, y):
    return divide(x, y)

def f(x, y):
    return call_divide(x, y)

f(1, 0) # sbb bagi input (1,0) , so x boleh bahagi 0

ZeroDivisionError: division by zero

**Question:** What has caused the Error?
<br><br><br><br><br><br>
**Answer:**cannot divide by 0

### Try except

Sometimes exceptions are expected to happen and we want our code to handle those exceptions in a certain way. For this there is the try except statement.

Let's try these inputs.

- typing a number
- typing a non-numeric string
- typing a zero
- ending the cell by intterupting the kernel

In [82]:
print(var) # dpt nameError sbb 'var' not declared

NameError: name 'var' is not defined

In [81]:
try:
    print(var)  # nak lari dari error tu, xyah run code ni
except:
    print("An exception occurred")

An exception occurred


In [84]:
x = [1,3,6,4,7,"6", 9,"5"]  # sbb ada string
print(sum(x))

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [83]:
x = [1,3,6,4,7,"6", 9,"5"]

try:
    print(sum(x))
except:
     print("There was An exception.")

There was An exception.


In [87]:
try:
    x = float(input("Please enter a number: "))
    print("inverse is:",  1/x)
except:
    print("Oops!  That was no valid number.")

Please enter a number: 3
inverse is: 0.3333333333333333


The try statement.

1. the __try clause__ (block under try:) is executed
1. if __no exception__ occurs, __except clause is skipped__
1. if __an exception occurs__, the rest of the try clause is skipped,
    - and __the first except__ clause matching the exception __is executed__,
    - if __no handler__ is found, execution stops, an error __Traceback is displayed__.
    
In this case we have used a bare except statement. This means all exception are catched including the KeyboardInterrupt.

You can specifiy which exception you want to be catched and you want to handle them.

Let's try typing an number, a non-numeric number, a zero and a keyboardintterup again.

In [ ]:
try:
    x = float(input("Please enter a number: "))
    print("inverse is:",  1/x)
except ValueError:
    print("Oops!  That was no valid number.  Try again...")
except ZeroDivisionError:
    print("Oops! Cannot divide by 0!")
except:
    print("Something else went wrong")
    # raise # return error message

else and finally are usefull to define clean-up action.
- ```else``` statements are executed __only if__ no exceptions occur in ```try``` block.
- ```finally``` statements are __always__ executed.

else clause avoids accidentally catching an exception.

In [ ]:
def divide(x, y):
    try:
        result = x / y
    except ZeroDivisionError:
        print("division by zero!")
    else:
        print("result is", result)
    finally:
        print("executing finally clause")
        
divide(2,0)

In [ ]:
divide(2,1)

In [ ]:
divide("a","b")

Let's do another case.

In [ ]:
total = 0
numbers = [1,2,3,"a", "b", "c", 5, 6, 7]

for number in numbers:
    total += number
    
total

We cannot add up strings to a number.

If we want to have the total of the numbers in the list we can ignore the strings.

In [89]:
total = 0
numbers = [1,2,3,"a", "b", "c", 5, 6, 7]

for number in numbers:
    try:
        total += number
    except:
        print('Something went wrong')
total

Something went wrong
Something went wrong
Something went wrong


24

In [90]:
total = 0
numbers = [1,2,3,"a", "b", "c", 5, 6, 7]
# kita tau, kalau org letak a,b,c dlam tu jadi typeError
for number in numbers:
    try:
        total += number
    except TypeError as e:
        print('Error: {}'.format(e)) # bagitau error apa yg kluar
total

Error: unsupported operand type(s) for +=: 'int' and 'str'
Error: unsupported operand type(s) for +=: 'int' and 'str'
Error: unsupported operand type(s) for +=: 'int' and 'str'


24

You can use the bare except statement. As long as you are mindful about that it catches all exceptions it's fine. For a script you operate yourself it can be fine.

In [91]:
total = 0
numbers = [1,2,3,"a", "b", "c", 5, 6, 7]

for number in numbers:
    try:
        total += number
    except:
        pass
total # xyah bagitau apa error dia, pass je dari a,b,c tu

24

### Summary

> #### try except
If an error is encountered, a try block code execution is stopped and transferred
down to the except block. 
```python
try:
    total += number
except:
    pass
```

> #### try except else finally
In addition to using an except block after the try block, you can also use the
finally block. The code in the finally block will be executed regardless of whether an exception
occurs. The else code is executed in case the try statement was a succes.
```python
try:
    result = x / y
except ZeroDivisionError:
    print("division by zero!")
else:
    print("result is", result)
finally:
    print("executing finally clause")
```

### RUN ME

Please run the below code snippet. It is required for running tests for your solution.

In [92]:
def test(got, expected):
    if got == expected:
        prefix = ' OK '
    else:
        prefix = '  FAIL '
    print((f' {prefix} got: {got} expected: {expected}'))

In [ ]:
test('a', 'ab')
test('a', 'a')

### Exercises

#### Exercise 1 - Fix it Multiply

We want to multiple the values of a and b of the dictionary. 

Can you help?

In [93]:
data = {"a": 10, "b": 20}

answer = data[a] * dala["b"] 

# TEST
print("fix_it_1")
test(answer, 200)

NameError: name 'a' is not defined

In [95]:
data = {"a": 10, "b": 20}

answer = data['a'] * data["b"] # value inside 'a' * value inside 'b'

# TEST
print("fix_it_1")
test(answer, 200)

fix_it_1
  OK  got: 200 expected: 200


#### Exercise 2 - Fix it Numbers

We want to extend numbers with another list and then get the sum. 

Can you help?

In [99]:
numbers = [1,2,3,4]
numbers = numbers.extend([5,6,7,8])

answer = sum(numbers)

# TEST
print("fix_it_numbers")
test(answer, 36)

fix_it_numbers
  OK  got: 36 expected: 36


In [ ]:
# Answer
numbers = [1,2,3,4]
numbers.extend([5,6,7,8]) # xyah letak variable

answer = sum(numbers)

# TEST
print("fix_it_numbers")
test(answer, 36)

#### Exercise 3 - Fix it Open

We try to read the contents of data/names_raw.txt into a string.

Can you help?

In [ ]:
import os

In [107]:
with open(os.path.join('..','data', "names_raw.txt"), 'r') as f:
    content = f.read()
    
# TEST
print("fix_it_open")
test(content, 'The names are:\n\nJeremy\nJan\nAkmal')

fix_it_open
   FAIL  got: The names are:

Jeremy
Narjes
Amin expected: The names are:

Jeremy
Jan
Akmal


#### Exercise 4 - Try salaries

Have a look at the data/salaries.txt file

Load it into a list of dictionaries and use try catch to handle the non-dictionary lines.

In [ ]:
salaries = ## your code
        
# TEST
print("try_salaries")
from answers import the_salaries
test(salaries, the_salaries)

In [102]:
from answers import the_salaries
the_salaries

[{'name': 'John', 'salary': 5000, 'years_employed': 4},
 {'name': 'Lee', 'salary': 4000, 'years_employed': 3},
 {'name': 'Alex', 'salary': 3000, 'years_employed': 7},
 {'name': 'John', 'salary': 3000, 'years_employed': 8},
 {'name': 'Lauren', 'salary': 6000, 'years_employed': 5}]

In [106]:
salaries = []

with open(os.path.join('..','data','salaries.txt')) as f:
    for line in f:
        try:
            salaries.append(json.loads(line))
        except: # kita nak tukar json jadi dict, kalau ada error, abaikan(pass)
                pass # error sbb format lain so x leh nak convert jadi dict
            
print("try_salaries")
from answers import the_salaries
test(salaries, the_salaries)

try_salaries
  OK  got: [{'name': 'John', 'salary': 5000, 'years_employed': 4}, {'name': 'Lee', 'salary': 4000, 'years_employed': 3}, {'name': 'Alex', 'salary': 3000, 'years_employed': 7}, {'name': 'John', 'salary': 3000, 'years_employed': 8}, {'name': 'Lauren', 'salary': 6000, 'years_employed': 5}] expected: [{'name': 'John', 'salary': 5000, 'years_employed': 4}, {'name': 'Lee', 'salary': 4000, 'years_employed': 3}, {'name': 'Alex', 'salary': 3000, 'years_employed': 7}, {'name': 'John', 'salary': 3000, 'years_employed': 8}, {'name': 'Lauren', 'salary': 6000, 'years_employed': 5}]


## Object Oriented Programming

Object-oriented Programming, or OOP for short, is a programming paradigm in which properties and behaviours are bundled into objects.

We will define a class Student. A class is a blue-print for an object. After defining the class Student there is only one Student class and you can create many Student objects out of that class.

Then we create a child class PythonStudent. The child class inherits all the properties from its parent and has it's own functionalities additionally.

### How to create a class

In [109]:
class Student: # --> parent class
# child class akan inherit properties of parent class + functionalities sendiri
    
    def __init__(self, name, grades=[]): # semua object mesti ada benda ni
        self.name = name
        self.grades = grades

### Instantiating objects


In [110]:
jeremy = Student("Jeremy", [10,4,5]) # instantiate object named 'jeremy'
# jeremy is the instance of the class 'Student'

In [111]:
print(jeremy.grades)
print(jeremy.name)

[10, 4, 5]
Jeremy


### Define methods in a class
We will extend this class with an average methods the returns the average grade of the student.

You can copy paste again and add the average function.

The repr() function returns a printable representation of the given object. 
repr is a special method used to represent a class's objects as a string. 
repr is called by the repr() built-in function.
You can define your own string representation of your class objects using the repr method.
Special methods are a set of predefined methods used to enrich your classes.


In [112]:
class Student: # --> parent class
# child class akan inherit properties of parent class + functionalities sendiri
    
    def __init__(self, name, grades=[]): # semua object mesti ada benda ni
        self.name = name
        self.grades = grades
        
    def __repr__(self):
        return "{}:{}".format(self.name, self.grades) # define class as string
    
    def average(self):
        return round(sum(self.grades) / len(self.grades), 2)


In [113]:
jeremy = Student("Jeremy", [10,4,5])
jeremy.average()

6.33

In OOP there is the concept of heritance. You can make a class that is the child of another class.

We will define a class PythonStudent.

In [117]:
# child class 'PythonStudent' belong to the parent class 'Student'
class PythonStudent(Student):
    def can_program(self): # a function, kalau true, then dia PythonStudent 
        return True

In [118]:
narjes = PythonStudent("Narjes", [10, 20]) # object 2 still bwk attribute parent class
narjes

Narjes:[10, 20]

In [119]:
narjes.average()

15.0

In [120]:
narjes.can_program() # dia instance kat child class

True

In [121]:
jeremy.can_program()
# jeremy bukan dalam child class

AttributeError: 'Student' object has no attribute 'can_program'

In [122]:
amin = PythonStudent("Amin", [30,20,15]) # object dalam child class

In [123]:
students = [jeremy, narjes, amin] # diorang semua dalam parent class
students

[Jeremy:[10, 4, 5], Narjes:[10, 20], Amin:[30, 20, 15]]

We can filter the Python students.

In [124]:
[n for n in students if isinstance(n, PythonStudent)] # nak check PythonStudent

[Narjes:[10, 20], Amin:[30, 20, 15]]

Well done!

### Summary

> #### class 
Python is an Object Oriented Programming language (OOP). This means that almost all the code is implemented using a special construct called classes. Programmers use classes to keep related things together. When defining a child class it will inherit the methods from the parent class. 

```python
class Student:
    
    def __init__(self, name, grades=[]):
        self.name = name
        self.grades = grades
      
    def average(self):
        return round(sum(self.grades) / len(self.grades), 2)
    

class PythonStudent(Student):
    
    def can_program(self):
        return True
```

### RUN ME

Please run the below code snippet. It is required for running tests for your solution.

In [ ]:
def test(got, expected):
    if got == expected:
        prefix = ' OK '
    else:
        prefix = '  FAIL '
    print((f' {prefix} got: {got} expected: {expected}'))

In [ ]:
test('a', 'ab')
test('a', 'a')

### Exercises

The CEO of your company happens to have a kid who is totally into Pokemon! For this reason he is very interested in finding out more about these Pokemon for himself. The thing is.. all these Pokemons are hidden away in files and spread out over different continents and countries folders. Oh no, what a bummer.

Ambitious as you are and on the verge of locking down that next promotion you walk in and say: 

> _"Calm down, I have just finished my Python-Fundamentals course, let me handle this."_

Use your Python skills in browsing folder structure, lists, dictionaries, reading from csv and writing to Excel to provide those insights and totally save the day!

Have a look the the pokeworld folder inside the data folder.

The folder structure is like this: continent/country/pokemon.txt

Each file represents a pokemon and inside the file is the description.

#### Exercise 1 - Get pokemon info

Get the continent, country, name and description for the pokemon in pokeworld/Asia/Malaysia/Wooper.txt

Hints: help(str.split), with open()

In [125]:
import os
pokefile = os.path.join('..,'"data", "pokeworld/Asia/Malaysia/Wooper.txt")

#### Exercise 2 - Create a pokemon class

Create a pokemon class which has three methods:
- \_\_init\_\_(self, pokefile) -- assign self.name, self.continent, self.country and self.description with contents from the file
- \_\_repr\_\_(self) -- represent a pokemon like 'Wooper from Malaysia, Asia'
- is_big_data(self) -- if there is 'Apache' or 'Hadoop' in self.description return True

In [ ]:
## your code here
    
# TEST
print("create_pokemon_class")
pokefile = os.path.join("data", "pokeworld/Asia/Malaysia/Wooper.txt")
pokefile2 = os.path.join("data", "pokeworld/Asia/Thailand/Hive.txt")

wooper = Pokemon(pokefile)
hive = Pokemon(pokefile2)

test(wooper.__repr__(), 'Wooper from Malaysia, Asia')
test(wooper.is_big_data(), False)
test(hive.is_big_data(), True)

#### Exercise 3 -  Load all pokemons

Now that we have defined the Pokemon class we can make a list of all the pokemon.

Hints: glob.glob, Pokemon(pokefile)

In [ ]:
## your code here

pokemons = ## your code here

# TEST
print("load_all_pokemons")
test(len(pokemons), 750)
test(type(pokemons[0]), Pokemon)

#### Exercise 4 - Provide insights

Now that we have a list of pokemon, let's answer some questions!

1. In which country is Pikachu?
2. How many pokemons in Malaysia?
3. How many pokemons in Asia?
4. Which continent has the most pokemons?
5. Which country has least pokemons?
6. Which countries have bigdata?

4.1 In which country is Pikachu?

4.2 How many pokemons in Malaysia?

4.3 How many pokemons in Asia?

4.4 Which continent has the most pokemons?

Hints: 
- make a dictionary continents for the count
- continents.setdefault
- sorted(continens.items(), key=... , reverse=True)

4.5 Which country has the least pokemon?

4.6 Which countries have big data?

Hints: set() to get unique values of a list.

# Well done! Congrats on completing Python Fundamentals!